### PR TITLE
Migration - Self Signup Config

### PR DESCRIPTION
**Pull request by thisaltennakoon**
## [Documentation](https://docs.google.com/document/d/17R7QCYeayIwFAmyB2jo-_4uCVPBVUlvFJK_uJJl5fUY/edit#heading=h.8egjgcg5e0ro)

### CODE REVIEW

**Comment by thisaltennakoon** [comment]
Please review @dushaniw @msm1992 @tharikaGitHub @tharindu1st

File: components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/client/V420Migration.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package org.wso2.carbon.apimgt.migration.migrator.client;
```
Comment (dushaniw) [inline components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/client/V420Migration.java:1]: license header missing

File: components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
+package org.wso2.carbon.apimgt.migration.migrator.v420;
```
Comment (dushaniw) [inline components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java:1]: license header missing

File: components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                UserRegistry registry = ServiceHolder.getRegistryService().getGovernanceSystemRegistry(tenantId);
+
+                log.info("WSO2 API-M Migration Task : Starting data migration of Self Signup Configuration for tenant "
+                        + tenantId + '(' + tenantDomain + ')');
+                if (registry.resourceExists("/apimgt/applicationdata/sign-up-config.xml")) {
+                    HashSet<String> signUpRoles = new HashSet<String>();
+                    String currentConfig = ServiceReferenceHolder.getInstance().getApimConfigService()
+                            .getTenantConfig(tenantDomain);
+                    JsonObject currentConfigJsonObject = (JsonObject) new JsonParser().parse(currentConfig);
+                    if (currentConfigJsonObject.has("SelfSignUp")) {
```
Comment (dushaniw) [inline components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java:72]: lets use a constant for the `SelfSignUp` role name

File: components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                                roleElement.getFirstChildWithName(new QName("IsExternalRole")).getText());
+                        String roleName = roleElement.getFirstChildWithName(new QName("RoleName")).getText();
+                        if (isExternalRole) {
+                            signUpRoles.add(signUpDomain + "/" + roleName);
+                        } else {
+                            signUpRoles.add("Internal/" + roleName);
+                        }
+                    }
+
+                    selfSignUpJsonObject.add("SignUpRoles", new Gson().toJsonTree(signUpRoles).getAsJsonArray());
```
Comment (dushaniw) [inline components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java:101]: lets use constant for `SignUpRoles`

File: components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                    // Prettify the tenant-conf
+                    Gson gson = new GsonBuilder().setPrettyPrinting().create();
+                    String formattedTenantConf = gson.toJson(currentConfigJsonObject);
+                    ServiceReferenceHolder.getInstance().getApimConfigService()
+                            .updateTenantConfig(tenantDomain, formattedTenantConf);
+
+                    registry.delete("/apimgt/applicationdata/sign-up-config.xml");
+                }
+                log.info("WSO2 API-M Migration Task : Completed data migration of Self Signup Configuration for tenant "
+                        + tenantId + '(' + tenantDomain + ')');
```
Comment (dushaniw) [inline components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java:113]: @thisaltennakoon shall we change the error handling logic as, if migration fails for one tenant, log the error and move to other tenant. Finally after all the tenants are iterated, throw the exception fail the migration.

File: components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                    }
+
+                    selfSignUpJsonObject.add("SignUpRoles", new Gson().toJsonTree(signUpRoles).getAsJsonArray());
+                    currentConfigJsonObject.add("SelfSignUp", selfSignUpJsonObject);
+
+                    // Prettify the tenant-conf
+                    Gson gson = new GsonBuilder().setPrettyPrinting().create();
+                    String formattedTenantConf = gson.toJson(currentConfigJsonObject);
+                    ServiceReferenceHolder.getInstance().getApimConfigService()
+                            .updateTenantConfig(tenantDomain, formattedTenantConf);
```
Comment (dushaniw) [inline components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java:108]: shall we log the migrated tenant-conf json config

File: components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(tenantId, true);
+                ServiceHolder.getTenantRegLoader().loadTenantRegistry(tenantId);
+                UserRegistry registry = ServiceHolder.getRegistryService().getGovernanceSystemRegistry(tenantId);
+
+                log.info("WSO2 API-M Migration Task : Starting data migration of Self Signup Configuration for tenant "
+                        + tenantId + '(' + tenantDomain + ')');
+                if (registry.resourceExists("/apimgt/applicationdata/sign-up-config.xml")) {
+                    HashSet<String> signUpRoles = new HashSet<String>();
+                    String currentConfig = ServiceReferenceHolder.getInstance().getApimConfigService()
```
Comment (dushaniw) [inline components/migration-client/wso2-api-migration-client/src/main/java/org/wso2/carbon/apimgt/migration/migrator/v420/V420RegistryResourceMigrator.java:69]: shall we log the current config before migration

**Review by dushaniw — APPROVED** [review]

**Pull request merged**